agents/acu/acu_agent.py

```diff
@@ -892,10 +892,14 @@ def _run_specified_scan(self, session, times, azs, els, vas, ves, azflags, elfla
                               }
                 self.agent.publish_to_feed('acu_upload', acu_upload, from_reactor=True)
             text = ''.join(upload_lines)
+            yield dsleep(0.05)
             free_positions = self.data['status']['summary']['Free_upload_positions']
             while free_positions < 9899:
                 free_positions = self.data['status']['summary']['Free_upload_positions']
-                yield dsleep(0.1)
+                yield dsleep(0.05)
+                if free_positions > 9997:
+                    self.log.warn('Scan aborted!')
+                    return False
```

Comment:
In the next loop down, the trap is on the axis mode (modes != ProgramTrack), though it is interpreted as a fault.  It would perhaps be better to use that mode check here, too.  If you are trying to discriminate between an unexpected fault (where free positions is not reset) and an abort (where free_positions -> 0), then that's great but you should probably do both things in both places.

I think the free_upload_positions check is not obviously reliable, because of the race condition between "stop_and_clear", which should set it to 10000, and this loop which is trying to upload points (which would usually put it below 9899).  It's not obvious that you couldn't get the sequence:
- _run_specified_scan: uploads some points ...
- _run_specified_scan: while free_pos < 9899, watch for abort condition...
- _run_specified_scan: prepare for next iteration
- _abort! stop_and_clear sets free_pos = 10000 and mode='stop'
- _run_specified_scan: uploads 120 points, free_pos -> 9700 or something
- _run_specified_scan: while free_pos < 9899, watch for abort condition <--- never exits, because points aren't being consumed.

Maybe that can't actually happen?  But it's not obvious, and we prefer obvious things.  And when something is true but not obvious it requires some comments!

So I think mode will be better.  Mode probably has its own weirdnesses, but at least we're not actively setting it in these loops.

Reply:
Also:

There are two loops, below, with dsleeps in them, that could in principle hang if the telescope suddenly stops moving.  Those need a similar check / checks.  (A local function to do this repeated task is probably a good idea.)

While you're in there I am tempted to request that all these stack size integers (10000, 9999, 9997 ... why 9997? 9899) get turned into constants / variables, or at least be explained... but I recognize they are a pre-existing condition.